rflx/pyrflx/typevalue.py

```diff
@@ -922,57 +931,50 @@ def bytestring(self) -> bytes:
     def fields(self) -> List[str]:
         return [f.name for f in self._type.fields]
 
-    def _update_accessible_fields(self) -> None:
-        nxt = self._next_field(INITIAL.name)
-        fields: List[str] = []
-        while nxt and nxt != FINAL.name:
-
-            if (
-                self.__simplified(self._type.field_condition(Field(nxt))) != TRUE
-                or not self._has_first(nxt)
-                or (
-                    not self._has_length(nxt)
-                    if not isinstance(self._fields[nxt].typeval, OpaqueValue)
-                    else not self._is_valid_opaque_field(nxt)
-                )
-            ):
-                break
-
-            fields.append(nxt)
-            nxt = self._next_field(nxt)
-        self.accessible_fields = fields
-
     def _is_valid_opaque_field(self, field: str) -> bool:
-        if self._get_length_unchecked(field) == UNDEFINED:
-            return False
 
-        for edge in self._type.incoming(Field(field)):
+        typeval = self._fields[field].typeval
+        assert isinstance(typeval, (ScalarValue, CompositeValue))
+        incoming = self._type.incoming(Field(field))
+        if isinstance(typeval, CompositeValue):
+            for l in incoming:
+                if (
+                    self.__simplified(l.condition) == TRUE
+                    and l.length != UNDEFINED
+                    and self._fields[l.source.name].set
+                ):
+                    break
+            else:
+                return False
+
+        for edge in incoming:
             if self.__simplified(edge.condition) == TRUE:
                 valid_edge = edge
                 break
         else:
             return True
 
         return all(
-            [
-                (v.name in self._fields and self._fields[v.name].set) or v.name == "Message"
-                for v in valid_edge.length.variables()
-            ]
+            (v.name in self._fields and self._fields[v.name].set) or v.name == "Message"
+            for v in valid_edge.length.variables()
         )
 
     @property
     def valid_fields(self) -> List[str]:
+        # ISSUE: nedbat/coveragepy#515
         return [
             f
             for f in self.accessible_fields
             if (
                 self._fields[f].set
                 and self.__simplified(self._type.field_condition(Field(f))) == TRUE
                 and any(
-                    [self.__simplified(i.condition) == TRUE for i in self._type.incoming(Field(f))]
+                    self.__simplified(i.condition) == TRUE
+                    for i in self._type.incoming(Field(f))  # pragma: no cover
```

Comment:
It seems to have something to do with https://github.com/nedbat/coveragepy/issues/515. Even for tests with only one element in incoming or outgoing I got the error that the generator expression was never fully run. I suspect that `any` aborts the calculation as soon as an element is true but the coverage analysis does not detect that there wouldn't be any further element in the generator.

Reply:
After some thinking the generator would run completely if none of the elements it produces is true. However this would require a field that is set and has no valid incoming nor a valid outgoing edge. This case is not possible with a valid model since a field without a valid incoming edge cannot be reached and therefore cannot be set and a field that is set but has no valid outgoing edge does violate the model.

Reply:
The `pragma: no cover` shouldn't be used in this way. It disables the coverage analysis for the whole return statement and not just for the generator expression.

> This case is not possible with a valid model

Couldn't we just remove the generator expression then?

Reply:
> Couldn't we just remove the generator expression then?

We could try to rewrite the list comprehension as a loop. I don't see any alternative to a generator expression otherwise.

Reply:
Why is removing the `any(...)` expression not an alternative? If it can never be false, why should it be checked at all?

Reply:
I tried to remove them but this made the test fail which checks the valid fields. So I assume there are cases where these checks are needed. However I still can't tell why an incomplete coverage is reported.

Reply:
Right. For the outgoing edges it should be possible to create a unit test, so that `pragma: no cover` is not needed anymore.

Reply:
Fixed in https://github.com/Componolit/RecordFlux/pull/381/commits/43cb59e028e040be26d5e714d897852fde513ae0.